### PR TITLE
fix: Reviewer's Assessment Activity must not be displayed in the Activity Flow Builder (M2-6852)

### PR DESCRIPTION
**Pull request by IhorNesterenko99**
- [x] Tests for the changes have been added
- [x] Delivered the `fix` or `feature` branches into `develop` or `release` branches via `Squash and Merge` (to keep clean history)

### 📝 Description

🔗 [Jira Ticket M2-6852](https://mindlogger.atlassian.net/browse/M2-6852)

Changes include:

- Clearing the Activities list of isReviewable Activities in the Activity Flow Builder.
- Clearing the Activities list of isReviewable Activities while creating the Activity Flow.
- Removing isReviewable Activities from Activity Flows when the isReviewable checkbox is checked.
- Removing isReviewable Activities from previously saved Activity Flows if they were saved with isReviewable Activities.
- Added unit tests.

### CODE REVIEW

**Comment by alex-i-e** [comment]
Could you please clarify what activities were in Flow_1 (combine reports) before change Assessment Activity checkbox?

After change that checkbox in that flow we got Activity_1 and Activity_1 which looks strange:
image

**Comment by IhorNesterenko99** [comment]
@alex-i-e 
We had Activity__2 in the Flow, which was removed after checking the checkbox. And yes, it can be strange because there's no limit on how many times the same Activity can be added to the same Flow.

File: src/modules/Builder/features/ActivityAbout/ActivityAbout.tsx [hunk 1]
```diff
@@ -95,6 +98,15 @@ export const ActivityAbout = () => {
     },
   ];
 
+  const handleIsReviewableChange = (event: ChangeEvent<HTMLInputElement>) => {
+    if (!event.target.checked) return;
+
+    const activityKey = getEntityKey(activity || {});
```
Comment (alex-i-e) [inline src/modules/Builder/features/ActivityAbout/ActivityAbout.tsx]: maybe it worth to add early return before it:
```
if (!activity) return; // or `setValue('activityFlows', activityFlows);` - if needed
```
```suggestion
    const activityKey = getEntityKey(activity);
```

File: src/modules/Builder/utils/getUpdatedActivityFlows.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { ActivityFlowFormValues } from 'modules/Builder/types';
+
+export const getUpdatedActivityFlows = (
+  activityFlows: ActivityFlowFormValues[],
+  keyToRemove: string,
+) =>
+  activityFlows.reduce((acc: ActivityFlowFormValues[], flow) => {
```
Comment (alex-i-e) [inline src/modules/Builder/utils/getUpdatedActivityFlows.ts:7]: could you a bit simplified the code to keep the reference as it was:
`keyToRemove ? (...) : activityFlows`

File: src/modules/Builder/features/ActivityFlowBuilder/ActivityFlowBuilder.tsx [hunk 3]
```diff
@@ -52,7 +52,12 @@ export const ActivityFlowBuilder = () => {
     name: `${activityFlowName}.items`,
     keyName: REACT_HOOK_FORM_KEY_NAME,
   });
-  const activities: AppletFormValues['activities'] = watch('activities');
+  const formActivities: AppletFormValues['activities'] = watch('activities');
```
Comment (alex-i-e) [inline src/modules/Builder/features/ActivityFlowBuilder/ActivityFlowBuilder.tsx]: here and in other places, please consider to use [useWatch](https://react-hook-form.com/docs/usewatch) :
```suggestion
  const formActivities: AppletFormValues['activities'] = useWatch({ name: 'activities'});
```
watch - This API will trigger re-render at the root of your app or form, consider using a callback or the useWatch api if you are experiencing performance issues.

File: src/modules/Builder/features/ActivityFlowBuilder/ActivityFlowBuilder.test.tsx [hunk 4]
```diff
@@ -202,4 +216,33 @@ describe('Activity Flow Builder', () => {
...
   });
+
+  test('Ensures no reviewable activities remain when adding/replacing an activity)', async () => {
+    renderActivityFlowBuilder(mockedAppletFormDataWithReviewableActivity);
+
+    const addActivity = screen.getByTestId(`${mockedActivityFlowBuilderTestid}-add`);
+    await userEvent.click(addActivity);
+
+    const activitiesInAddActivity = screen.getAllByTestId(
+      new RegExp(`^${mockedActivityFlowBuilderTestid}-add-activity-\\d+$`),
```
Comment (alex-i-e) [inline src/modules/Builder/features/ActivityFlowBuilder/ActivityFlowBuilder.test.tsx]: Let's move `new RegExp(`^${mockedActivityFlowBuilderTestid}-add-activity-\\d+$`)` into a constant here and in row 243 to reduce new object creation in the loops.

File: src/modules/Builder/utils/getUpdatedActivityFlows.ts (hunk 2 above)
Comment (IhorNesterenko99) [reply src/modules/Builder/utils/getUpdatedActivityFlows.ts:7]: Could you please provide more details?

**Review by alex-i-e — APPROVED** [review]

**Pull request merged**